### PR TITLE
llvmPackages_9: remove due to age

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Description of changes

Part of #283954

CC @wegank 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Follow up from #284865

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by annafleed** [comment]
can remove this too:
https://github.com/NixOS/nixpkgs/blob/34852ddf5e9203075805af70c3e556292c03ddb3/pkgs/build-support/cc-wrapper/default.nix#L565-L569

and should add aliases and an entry in pkgs/test/default.nix. see:
https://github.com/NixOS/nixpkgs/pull/283591/files#diff-308706790568e4d1381f5ab493c6eed2c30917e88ec06ae08d4ba9f049e8b64a

**Comment by RossComputerGuy** [comment]
> can remove this too:
> 
> https://github.com/NixOS/nixpkgs/blob/34852ddf5e9203075805af70c3e556292c03ddb3/pkgs/build-support/cc-wrapper/default.nix#L565-L569
> 
> and should add aliases and an entry in pkgs/test/default.nix. see: https://github.com/NixOS/nixpkgs/pull/283591/files#diff-308706790568e4d1381f5ab493c6eed2c30917e88ec06ae08d4ba9f049e8b64a

Looks good now?

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -19272,8 +19256,8 @@ with pkgs;
   lttv = callPackage ../development/tools/misc/lttv { };
 
   luaformatter = callPackage ../development/tools/luaformatter
-    (lib.optionalAttrs (stdenv.cc.isClang && lib.versionOlder stdenv.cc.version "9") {
-      stdenv = overrideCC stdenv llvmPackages_9.clang;
+    (lib.optionalAttrs (stdenv.cc.isClang && lib.versionOlder stdenv.cc.version "12") {
+      stdenv = overrideCC stdenv llvmPackages_12.clang;
```
Comment (annafleed) [inline pkgs/top-level/all-packages.nix:19260]: i think this can just be removed

Comment (RossComputerGuy) [reply pkgs/top-level/all-packages.nix]: Alright, done

**Review by ghost — COMMENTED** [review]
LGTM tho should verify xeus-cling builds

**Comment by RossComputerGuy** [comment]
> LGTM tho should verify xeus-cling builds

Yeah, I don't know anything about xeus-cling so someone who does know about it should look into it.

**Comment by wegank** [comment]
`xeus-cling` isn't compatible with recent versions of LLVM yet, so probably should be marked as broken. Or there's a simple fix:

```
  postPatch = ''
    substituteInPlace src/xmagics/executable.cpp \
      --replace "getDataLayout" "getDataLayoutString"
    substituteInPlace src/xmagics/execution.cpp \
      --replace "simplisticCastAs" "castAs"
    substituteInPlace src/xmime_internal.hpp \
      --replace "code.str()" "code.str().str()"
  '';
```

**Comment by RossComputerGuy** [comment]
> `xeus-cling` isn't compatible with recent versions of LLVM yet, so probably should be marked as broken. Or there's a simple fix, from [jupyter-xeus/xeus-cling#513](https://github.com/jupyter-xeus/xeus-cling/issues/513):
> 
> ```
>   postPatch = ''
>     substituteInPlace src/xmagics/executable.cpp \
>       --replace "getDataLayout" "getDataLayoutString"
>     substituteInPlace src/xmagics/execution.cpp \
>       --replace "simplisticCastAs" "castAs"
>     substituteInPlace src/xmime_internal.hpp \
>       --replace "code.str()" "code.str().str()"
>   '';
> ```

The patch just seems simpler.

**Comment by annafleed** [comment]
@ofborg build xeus-cling xeus-cling.passthru.tests

**Comment by annafleed** [comment]
woot!

File: pkgs/top-level/aliases.nix [hunk 2]
```diff
@@ -199,7 +199,7 @@ mapAliases ({
   clang6Stdenv = throw "clang6Stdenv has been removed from nixpkgs"; # Added 2024-01-08
   clang7Stdenv = throw "clang7Stdenv has been removed from nixpkgs"; # Added 2023-11-19
   clang8Stdenv = throw "clang8Stdenv has been removed from nixpkgs"; # Added 2024-01-24
-  clang9Stdenv = lowPrio llvmPackages_9.stdenv;
+  clang9Stdenv = throw "clang9Stdenv has been remvoed from nixpkgs"; # Added 2024-04-08
```
Comment (philiptaron) [inline pkgs/top-level/aliases.nix:202]: typo here

Comment (RossComputerGuy) [reply pkgs/top-level/aliases.nix:202]: Oh no

Comment (wegank) [reply pkgs/top-level/aliases.nix:202]: #303012

**Pull request merged**